### PR TITLE
Improve performance of InCodeGenerator

### PR DESCRIPTION
**Pull request by haozhun**
* Threshold between hash set and generated switch is adjusted from 1000 to 32.
* Load factor of hash set is adjusted to 0.25.

BIGINT is 31-bit unsigned integers. DOUBLE is any double between 0.0 and 1.0. VARCHAR is 64-bit signed integer converted to string. Unit is microsecond for a 10000-row page.

Overall performance after tuning (10 fork, 10 warmup, 10 run, Linux E5-2660):
```
            bigint             double             varchar
    1    71.299 ± 0.593     75.100 ± 0.250    392.109 ± 18.397
    5   139.101 ± 4.758    160.138 ± 7.903    488.277 ±  6.884
   10   171.374 ± 3.307    243.076 ± 3.173    542.460 ±  6.216
   25   216.118 ± 2.685    293.696 ± 3.388    602.593 ±  3.933
   50   226.435 ± 1.825    233.620 ± 2.641    586.078 ±  7.143
   75   208.593 ± 1.599    212.449 ± 0.936    557.064 ±  7.618
  100   226.122 ± 0.988    229.184 ± 1.151    592.266 ±  6.021
  150   206.488 ± 1.126    212.651 ± 1.396    570.001 ±  9.624
  200   227.632 ± 1.801    234.063 ± 2.066    591.663 ±  5.204
  250   248.663 ± 2.260    254.747 ± 1.793    624.223 ±  6.793
  300   209.389 ± 1.317    212.957 ± 1.327    567.363 ±  4.229
  350   215.955 ± 1.120    222.608 ± 1.559    584.569 ±  5.289
  400   228.595 ± 1.211    233.784 ± 1.366    596.820 ±  6.282
  450   239.963 ± 1.549    243.192 ± 1.362    620.790 ±  3.858
  500   250.862 ± 1.622    256.127 ± 2.767    634.933 ±  4.485
  750   223.741 ± 1.275    226.601 ± 1.218    595.052 ±  6.548
 1000   250.443 ± 0.982    254.663 ± 1.343    653.762 ±  4.397
10000   220.256 ± 1.061    226.338 ± 0.910    661.664 ±  7.526
```
Benchmark results for different strategies (1 fork, 5 warmup, 10 run, OSX i7-4960HQ):
```
        hash 0.75 load factor    hash 0.5 load factor     hash 0.25 load factor     generated switch
        BIGINT  DOUBLE  VARCHAR  BIGINT  DOUBLE  VARCHAR  BIGINT  DOUBLE  VARCHAR   BIGINT  DOUBLE   VARCHAR
    1   235.74  225.23   706.36  182.28  181.29   588.15  163.85  165.74  545.860    39.76    51.12   254.47
    5   388.62  305.17   999.64  232.17  221.40   656.43  174.95  164.54  580.087    89.19   119.79   367.25
   10   427.94  360.83  1122.43  223.10  230.58   684.98  175.28  171.46  600.098   139.09   198.23   396.69
   25   283.84  271.31   753.11  273.71  266.14   834.23  183.88  181.93  616.971   178.48   218.09   435.17
   50   268.04  270.22   817.26  267.91  250.73   798.80  202.50  184.64  649.919   249.21   254.60   454.71
   75   398.50  382.17  1306.62  232.38  233.09   650.71  168.15  164.87  568.743   262.46   278.25   484.06
  100   261.93  280.38   906.26  270.54  276.77   816.94  187.18  191.77  614.767   270.46   282.47   493.71
  150   424.19  395.80  1318.54  231.03  226.42   691.79  174.91  173.50  592.280   293.14   292.15   513.29
  200   287.61  263.83   838.44  280.64  282.62   826.32  193.01  190.98  623.896   305.62   313.68   519.65
  250   364.81  490.35  1038.71  372.73  334.62  1068.17  206.48  214.01  662.589   316.55   323.35   545.67
  300   422.06  385.28  1449.77  214.26  216.09   732.39  175.13  171.39  594.354   332.34  1750.36  1955.22
  350   536.78  554.96  1648.89  246.64  257.35   810.53  185.83  185.97  611.776   336.92  1774.38  2028.32
  400   277.07  281.35   860.19  280.14  276.10   910.70  193.19  190.66  625.258   349.76  1784.38  1987.22
  450   309.79  306.55  1019.18  304.69  304.47   956.27  203.24  198.90  660.130   352.02  1718.63  1976.81
  500   391.65  343.55  1050.55  335.05  340.80  1066.33  217.58  222.25  667.676   350.37  1719.94  1968.86
  750   719.82  737.52  2510.43  283.80  268.24   827.40  181.36  185.56  617.092   387.30  1827.96  2092.34
 1000   350.08  345.73  1060.12  338.92  344.01  1026.75  214.20  215.72  692.699  1769.19  1800.66  2045.39
10000   464.89  451.28  1786.75  244.11  237.86   750.99  184.81  183.78  659.567
```

### CODE REVIEW

File: presto-main/src/test/java/com/facebook/presto/sql/gen/InCodeGeneratorBenchmark.java [hunk 1]
```diff
@@ -0,0 +1,152 @@
...
+import static com.facebook.presto.sql.relational.Expressions.call;
+import static com.facebook.presto.sql.relational.Expressions.constant;
+import static com.facebook.presto.sql.relational.Expressions.field;
+import static com.facebook.presto.sql.relational.Signatures.IN;
+import static com.google.common.base.Preconditions.checkState;
+import static org.openjdk.jmh.annotations.Mode.AverageTime;
+
+@State(Scope.Thread)
+@OutputTimeUnit(TimeUnit.MICROSECONDS)
+@Fork(1)
```
Comment (martint) [inline presto-main/src/test/java/com/facebook/presto/sql/gen/InCodeGeneratorBenchmark.java]: You should run multiple forks (typically ~10) to account for run-to-run variance.

File: presto-main/src/test/java/com/facebook/presto/sql/gen/InCodeGeneratorBenchmark.java [hunk 2]
```diff
@@ -0,0 +1,152 @@
...
+@Warmup(iterations = 5, time = 500, timeUnit = TimeUnit.MILLISECONDS)
```
Comment (martint) [inline presto-main/src/test/java/com/facebook/presto/sql/gen/InCodeGeneratorBenchmark.java]: Is this enough to get a stabilized measurement?

File: presto-main/src/test/java/com/facebook/presto/sql/gen/InCodeGeneratorBenchmark.java [hunk 3]
```diff
@@ -0,0 +1,152 @@
...
+@Measurement(iterations = 10, time = 500, timeUnit = TimeUnit.MILLISECONDS)
+@BenchmarkMode(AverageTime)
+@SuppressWarnings("FieldMayBeFinal")
+public class InCodeGeneratorBenchmark
+{
+    @Param({"1", "5", "10", "25", "50", "75", "100", "150", "200", "250", "300", "350", "400", "450", "500", "750", "1000", "10000"})
+    private int inListCount = 1;
+
+    @Param({"BIGINT", "DOUBLE", "VARCHAR"})
```
Comment (martint) [inline presto-main/src/test/java/com/facebook/presto/sql/gen/InCodeGeneratorBenchmark.java]: Use the constants in `StandardTypes`

**Comment by haozhun** [comment]
https://gist.github.com/haozhun/83b5fa90f95c449dd316

**Comment by dain** [comment]
This needs release notes

**Comment by haozhun** [comment]
Addressed comments

File: presto-docs/src/main/sphinx/release/release-0.126.rst [hunk 4]
```diff
@@ -14,6 +14,8 @@ General Changes
   within a task. This allows us to optimize for index cache hits or for more
   CPU parallelism. This option is toggled by the ``task.share-index-loading``
   config property or the ``task_share_index_loading`` session property.
+* Improve performance of queries that use ``IN`` expressions with large lists
```
Comment (dain) [inline presto-docs/src/main/sphinx/release/release-0.126.rst]: `use an IN expression with a large list`

**Comment by dain** [comment]
looks good

**Comment by haozhun** [comment]
Merged #3953

**Comment by damiencarol** [comment]
Happy to see this one.
I have queries with larges _IN_ clauses ( < 30, > 1000 and about 200-500).
I will see the performance impact.

**Comment by haozhun** [comment]
Glad it helps. We saw unintuitive performance characteristics for 400-ish IN clauses. And that's why we decided to dig into this. This fix benefits IN clauses between 300-1000 most (up to 7X). It also tuned performance for others (20% to 200% speedup).

**Comment by haozhun** [comment]
Related to #3836.

**Pull request closed without merging**